File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
   public String smallestFromLeaf(TreeNode root) {
        return sol(root, "");
    }

    private String sol(TreeNode root, String ans) {
        if (root == null) return "";
        if (root.left == null && root.right == null) {
            ans = (char) (root.val + 'a') + ans;
            return ans;
        }
        ans = (char) (root.val + 'a') + ans;
        String left = sol(root.left, ans);
        String right = sol(root.right, ans);
        String a = (left.equals("")) ? right : left;
        String b = (right.equals("")) ? left : right;
        return a.compareTo(b) < 0 ? a : b;
    }
}

File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string sol(TreeNode* root,string ans){
        if(!root)return "";
        if(!root->left && !root->right){ans=char(root->val+'a')+ans;return ans;}
        ans=char(root->val+'a')+ans;
        string left=sol(root->left,ans);
        string right=sol(root->right,ans);
        string a=(left=="")?right:left;
        string b=(right=="")?left:right;
        return min(a,b);
    }
    string smallestFromLeaf(TreeNode* root) {
        return sol(root,"");
    }
};